### PR TITLE
Chore: Remove opentracing and use opentelemetry instead

### PR DESCRIPTION
**Pull request by zserge**
Parent Issue: https://github.com/grafana/grafana/issues/65793

This PR removes OpenTracing implementation from `pkg/infra/tracing` and uses OpenTelemetry instead.
Legacy config format is still supported, though.

Previously tests have only been testing the initialisation of the tracer, and those have been ported to the new implementation.

Proper "integration" tests that would ensure that tracer is _working_ are to be done yet.

### CODE REVIEW

File: pkg/infra/tracing/tracing_test.go [hunk 1]
```diff
@@ -5,137 +5,149 @@ import (
...
-	ts := &Opentracing{Cfg: cfg}
-	_, err := ts.Cfg.Raw.NewSection("tracing.jaeger")
-	require.NoError(t, err)
-	require.NoError(t, ts.parseSettings())
-	jaegerCfg, err := ts.initJaegerCfg()
-	require.NoError(t, err)
+func TestSplitCustomAttribs_Malformed(t *testing.T) {
+	tests := []struct {
+		input    string
+		expected []attribute.KeyValue
```
Comment (papagian) [inline pkg/infra/tracing/tracing_test.go:47]: is this used at all?

File: pkg/infra/tracing/tracing_test.go [hunk 2]
```diff
@@ -5,137 +5,149 @@ import (
...
+		Env                map[string]string
+		ExpectedExporter   string
+		ExpectedAddress    string
+		ExpectedPropagator string
+		ExpectedAttrs      []attribute.KeyValue
+	}{
+		{
+			Name:             "default config uses noop exporter",
+			Cfg:              "",
+			ExpectedExporter: "noop",
```
Comment (papagian) [inline pkg/infra/tracing/tracing_test.go:72]: nit, check for the [constant](https://github.com/grafana/grafana/blob/a68c9d0616a5b7b90fbf4274a20297ab7664659b/pkg/infra/tracing/tracing.go#L42) instead

File: pkg/infra/tracing/tracing_test.go [hunk 3]
```diff
@@ -5,137 +5,149 @@ import (
...
+		{
+			Name: "legacy config format is supproted",
+			Cfg: `
+			[tracing.jaeger]
+			address = jaeger.example.com:6831
+			`,
+			ExpectedExporter: "jaeger",
+			ExpectedAddress:  "jaeger.example.com:6831",
+			ExpectedAttrs:    []attribute.KeyValue{},
+		},
```
Comment (papagian) [inline pkg/infra/tracing/tracing_test.go:113]: I would worth having a test if both sections `tracing.jaeger` and `tracing.opentelemetry` exist.

File: pkg/infra/tracing/tracing_test.go [hunk 4]
```diff
@@ -5,137 +5,149 @@ import (
...
+			Cfg: `
+			[tracing.opentelemetry.otlp]
+			address = otlp.example.com:4317
+			`,
+			ExpectedExporter: "otlp",
+			ExpectedAddress:  "otlp.example.com:4317",
+			ExpectedAttrs:    []attribute.KeyValue{},
+		},
+		{
+			Name: "legacy config format is supproted",
```
Comment (papagian) [inline pkg/infra/tracing/tracing_test.go:105]:
```suggestion
			Name: "legacy config format is supported",
```

**Review by papagian — COMMENTED** [review]
Great work 🎉 
the changes look good AFAIKT: I have some minor suggestions
I will try to test manually a bit and I will provide my feedback soon

File: pkg/infra/tracing/tracing.go [hunk 5]
```diff
@@ -173,219 +194,343 @@ func ContextWithSpan(ctx context.Context, span Span) context.Context {
...
 
-	_, err := cfg.FromEnv()
+	return res, nil
+}
+
+func (ots *Opentelemetry) initJaegerTracerProvider() (*tracesdk.TracerProvider, error) {
+	var ep jaeger.EndpointOption
+	// Create the Jaeger exporter: address can be either agent address (host:port) or collector URL
+	if host, port, err := net.SplitHostPort(ots.Address); err == nil {
+		ep = jaeger.WithAgentEndpoint(jaeger.WithAgentHost(host), jaeger.WithAgentPort(port))
```
[the remark this answers is not in the document]
Comment (zserge) [reply pkg/infra/tracing/tracing.go:270]: Well, it's more of a combination: previously Jaeger seemed to use both but with OpenTelemetry we somehow only used the HTTP collector and not the UDP agent. Now we support both again.

File: pkg/infra/tracing/tracing.go [hunk 6]
```diff
@@ -173,219 +194,343 @@ func ContextWithSpan(ctx context.Context, span Span) context.Context {
...
+		ots.enabled = jaegerExporter
+		return nil
 	}
 
-	ts.customTags = splitTagSettings(section.Key("always_included_tag").MustString(""))
-	ts.samplerType = section.Key("sampler_type").MustString("")
-	ts.samplerParam = section.Key("sampler_param").MustFloat64(1)
-	ts.zipkinPropagation = section.Key("zipkin_propagation").MustBool(false)
-	ts.disableSharedZipkinSpans = section.Key("disable_shared_zipkin_spans").MustBool(false)
-	ts.samplingServerURL = section.Key("sampling_server_url").MustString("")
```
[the remark this answers is not in the document]
Comment (zserge) [reply pkg/infra/tracing/tracing.go:218]: Zipkin is definitely broken and has been for a while. I think we should re-introduce it as a separate tracer.
Sampling server URL is properly backported to OpenTelemetry tracer.

Sampling in general is something I'd rather remove at some point (Tempo seems to handle it just find and with Jaeger it's better to configure sampling on the agent side).

File: pkg/infra/tracing/tracing_test.go [hunk 7]
```diff
@@ -5,137 +5,149 @@ import (
...
+			expected: []attribute.KeyValue{
+				attribute.String("key1", "value1"),
+				attribute.String("key2", "value2"),
+			},
 		},
 		{
-			input:    "tag1",
-			expected: map[string]string{},
+			input:    "",
+			expected: []attribute.KeyValue{},
```
[the remark this answers is not in the document]
Comment (zserge) [reply pkg/infra/tracing/tracing_test.go:33]: Ehm.. The diff is a lie here, I moved OpenTelemetry test as-is and removed the OpenTracing one. Git somehow shows like if the test was edited. `tag1` is malformed input, which is tested separately in the following OpenTelemetry test.

**Review by papagian — APPROVED** [review]
I have tested the current branch using:
`JAEGER_AGENT_HOST=localhost JAEGER_AGENT_PORT=6831 GF_LOG_LEVEL=debug make run`

and

`GF_TRACING_JAEGER_ADDRESS=localhost:6831 GF_LOG_LEVEL=debug make run`

and

`GF_TRACING_OPENTELEMETRY_JAEGER_ADDRESS=localhost:6831 GF_LOG_LEVEL=debug make run`

and works as expected

**Pull request merged**